pkg/operation/botanist/dns.go

```diff
@@ -64,25 +69,119 @@ func (b *Botanist) DeployExternalDomainDNSRecord(ctx context.Context) error {
 		return nil
 	}
 
-	if err := b.deployDNSProvider(ctx, DNSPurposeExternal, b.Shoot.ExternalDomain.Provider, b.Shoot.ExternalDomain.SecretData, sets.NewString(append(b.Shoot.ExternalDomain.IncludeDomains, *b.Shoot.ExternalClusterDomain)...).List(), b.Shoot.ExternalDomain.ExcludeDomains, b.Shoot.ExternalDomain.IncludeZones, b.Shoot.ExternalDomain.ExcludeZones); err != nil {
+	if err := b.deployDNSProvider(ctx, DNSPurposeExternal, DNSProviderRolePrimary, b.Shoot.ExternalDomain.Provider, b.Shoot.ExternalDomain.SecretData, sets.NewString(append(b.Shoot.ExternalDomain.IncludeDomains, *b.Shoot.ExternalClusterDomain)...).List(), b.Shoot.ExternalDomain.ExcludeDomains, b.Shoot.ExternalDomain.IncludeZones, b.Shoot.ExternalDomain.ExcludeZones); err != nil {
 		return err
 	}
 	return b.deployDNSEntry(ctx, DNSPurposeExternal, common.GetAPIServerDomain(*b.Shoot.ExternalClusterDomain), common.GetAPIServerDomain(b.Shoot.InternalClusterDomain))
 }
 
 // DestroyExternalDomainDNSRecord destroys the DNS record for the external cluster domain.
 func (b *Botanist) DestroyExternalDomainDNSRecord(ctx context.Context) error {
-	if err := b.deleteDNSEntry(ctx, DNSPurposeExternal); err != nil {
+	return b.deleteDNSEntry(ctx, DNSPurposeExternal)
+}
+
+// DeployAdditionalDNSProviders deploys the additional DNS providers configured in the shoot resource.
+func (b *Botanist) DeployAdditionalDNSProviders(ctx context.Context) error {
+	if b.Shoot.Info.Spec.DNS == nil {
+		return nil
+	}
+
+	var (
+		fns               []flow.TaskFn
+		deployedProviders = sets.NewString()
+	)
+	for i, provider := range b.Shoot.Info.Spec.DNS.Providers {
+		p := provider
+		if p.Primary != nil && *p.Primary {
+			continue
+		}
+		fns = append(fns, func(ctx context.Context) error {
+			var includeDomains, excludeDomains, includeZones, excludeZones []string
+			if domains := p.Domains; domains != nil {
+				includeDomains = domains.Include
+				excludeDomains = domains.Exclude
+			}
+			if zones := p.Zones; zones != nil {
+				includeZones = zones.Include
+				excludeZones = zones.Exclude
+			}
+			providerType := p.Type
+			if providerType == nil {
+				return fmt.Errorf("dns provider[%d] doesn't speify a type", i)
+			}
+			if *providerType == core.DNSUnmanaged {
+				b.Logger.Infof("Skipping deployment of DNS provider[%d] since it specifies type %q", i, core.DNSUnmanaged)
+				return nil
+			}
+			secretName := p.SecretName
+			if secretName == nil {
+				return fmt.Errorf("dns provider[%d] doesn't specify a secretName", i)
+			}
+			secret := &corev1.Secret{}
+			if err := b.K8sGardenClient.Client().Get(ctx, kutil.Key(b.Shoot.Info.Namespace, *secretName), secret); err != nil {
+				return fmt.Errorf("could not get dns provider secret %q: %+v", *secretName, err)
+			}
```

Comment:
Can this entire thing moved into https://github.com/gardener/gardener/blob/b375d42f6006d863a2f5cf1ec835a46c23174ce1/pkg/operation/garden/types.go#L21-L26

With this fetching this would be need only once during a reconciliation flow

Reply:
I'm not sure why you propose to add it to the `Garden` struct since it only holds information related to the whole environment and nothing specific to the shoot.

Imo, additional DNS providers don't have to be part of the `Operation` struct either because the data is only needed in the deploy step or no other func of `Operation` needs to access that data as well. I know that this will change with your isolation plans but would still prefer to leave it like this for the moment.